File: Stack/84.largest-rectangle-in-histogram.go
package leetcode

// @leet start
func largestRectangleArea(heights []int) int {
	var (
		res   int
		st    []int
		n     = len(heights)
		left  = make([]int, n)
		right = make([]int, n)
	)
	// for i := range n {
	// 	left[i] = -1
	// }
	// for i := range n {
	// 	right[i] = n
	// }
	for i := range n {
		v := heights[i]
		for len(st) > 0 && heights[st[len(st)-1]] >= v {
			st = st[:len(st)-1]
		}
		if len(st) > 0 {
			left[i] = st[len(st)-1]
		} else {
			left[i] = -1
		}
		st = append(st, i)
	}

	st = st[:0]
	for i := n - 1; i >= 0; i-- {
		v := heights[i]
		for len(st) > 0 && heights[st[len(st)-1]] >= v {
			st = st[:len(st)-1]
		}
		if len(st) > 0 {
			right[i] = st[len(st)-1]
		} else {
			right[i] = n
		}
		st = append(st, i)
	}

	for i := range n {
		res = max(res, heights[i]*(right[i]-left[i]-1))
	}

	return res
}

// @leet end

File: Stack/84.largest-rectangle-in-histogram.py
from typing import List


# @leet start
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = [-1]
        max_area = 0

        for i, h in enumerate(heights):
            while stack[-1] != -1 and h <= heights[stack[-1]]:
                height = heights[stack.pop()]
                width = i - stack[-1] - 1
                max_area = max(max_area, height * width)
            stack.append(i)

        while stack[-1] != -1:
            height = heights[stack.pop()]
            width = len(heights) - stack[-1] - 1
            max_area = max(max_area, height * width)

        return max_area

    def largestRectangleArea2(self, heights: List[int]) -> int:
        n = len(heights)
        l_small = [-1] * n
        r_small = [n] * n
        ret = 0
        st = []
        for i in range(n):
            while st and heights[st[-1]] >= heights[i]:
                st.pop()
            if st:
                l_small[i] = st[-1]
            else:
                l_small[i] = -1
            st.append(i)

        st.clear()
        for i in range(n - 1, -1, -1):
            while st and heights[st[-1]] >= heights[i]:
                st.pop()
            if st:
                r_small[i] = st[-1]
            else:
                r_small[i] = n
            st.append(i)

        for i in range(n):
            r, l = r_small[i], l_small[i]
            ret = max(ret, heights[i] * (r - l - 1))
        return ret


# @leet end
